libs/oracledb/langchain_oracledb/vectorstores/oraclevs.py

```diff
@@ -0,0 +1,1929 @@
+        return await client.acquire()
+    else:
+        valid_types = "oracledb.AsyncConnection"
+        if connection_pool_class:
+            valid_types += " or oracledb.AsyncConnectionPool"
+        raise TypeError(
+            f"Expected client of type {valid_types}, got {type(client).__name__}"
+        )
+
+
+def _handle_exceptions(func: T) -> T:
+    @functools.wraps(func)
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
+        try:
+            return func(*args, **kwargs)
+        except RuntimeError as db_err:
+            # handle a known type of error (e.g., DB-related) specifically
+            logger.exception("DB-related error occurred.")
+            raise RuntimeError(
+                "Failed due to a DB issue: {}".format(db_err)
+            ) from db_err
+        except ValueError as val_err:
+            # handle another known type of error specifically
+            logger.exception("Validation error.")
+            raise ValueError("Validation failed: {}".format(val_err)) from val_err
+        except Exception as e:
+            # generic handler for all other exceptions
+            logger.exception("An unexpected error occurred: {}".format(e))
+            raise RuntimeError("Unexpected error: {}".format(e)) from e
+
+    return cast(T, wrapper)
+
+
+def _table_exists(connection: Connection, table_name: str) -> bool:
+    try:
+        with connection.cursor() as cursor:
+            cursor.execute(f"select 1 from {table_name} where ROWNUM < 1")
+            return True
+    except oracledb.DatabaseError as ex:
+        err_obj = ex.args
+        if err_obj[0].code == 942:
+            return False
+        raise
+
+
+async def _atable_exists(connection: AsyncConnection, table_name: str) -> bool:
+    try:
+        with connection.cursor() as cursor:
+            await cursor.execute(f"select 1 from {table_name} where ROWNUM < 1")
+            return True
+    except oracledb.DatabaseError as ex:
+        err_obj = ex.args
+        if err_obj[0].code == 942:
+            return False
+        raise
+
+
+def _compare_version(version: str, target_version: str) -> bool:
+    # split both version strings into parts
+    version_parts = [int(part) for part in version.split(".")]
+    target_parts = [int(part) for part in target_version.split(".")]
+
+    # compare each part
+    for v, t in zip(version_parts, target_parts):
+        if v < t:
+            return True  # current version is less
+        elif v > t:
+            return False  # current version is greater
+
+    # if all parts equal so far, check if version has fewer parts than target_version
+    return len(version_parts) < len(target_parts)
+
+
+@_handle_exceptions
+def _index_exists(connection: Connection, index_name: str) -> bool:
+    # check if the index exists
+    query = """
+        SELECT index_name 
+        FROM all_indexes 
+        WHERE upper(index_name) = upper(:idx_name)
```

Comment:
Does this cope with mixed case names?

Reply:
I agree that this check was not working very well. I changed to better handle mixed case. 
Also, his check would not let users create an index if there was an index with a same name in another table and do not give any errors or logs. I added table_name to the check, now users will get `oracledb.exceptions.DatabaseError: ORA-00955: name is already used by an existing object` when they try to create indexes for different tables with the same name. New tests added for it `test_index_table_case` and `test_index_table_case_async`